torch/optim/optimizer.py

```diff
@@ -137,17 +165,75 @@ def wrapper(*args, **kwargs):
                 obj, *_ = args
                 profile_name = "Optimizer.step#{}.step".format(obj.__class__.__name__)
                 with torch.autograd.profiler.record_function(profile_name):
+                    # call optimizer step pre hooks
+                    for pre_hook in self._optimizer_step_pre_hooks.values():
+                        result = pre_hook(self, args, kwargs)
+                        if result is not None:
+                            if isinstance(result, tuple) and len(result) == 2:
+                                args, kwargs = result
+                            else:
+                                raise RuntimeError(f"{func} must return None or a tuple of (new_args, new_kwargs),"
+                                                   f"but got {result}.")
+
                     out = func(*args, **kwargs)
                     obj._optimizer_step_code()
+
+                    # call optimizer step post hooks
+                    for post_hook in self._optimizer_step_post_hooks.values():
+                        post_hook(self, args, kwargs)
+
                     return out
 
             return wrapper
 
+        self._zero_grad_profile_name = "Optimizer.zero_grad#{}.zero_grad".format(self.__class__.__name__)
```

Comment:
to improve readability - all code of the parent function now is positioned together rather than spread out before and after the `profile_hook_step` function. I can revert it if you think the original presentation is easier to read.